limma-voom example for RNA-seq w/o quantile normalization (#95)
AlexsLemonade/refinebio-examples: differential-expression/rnaseq_DGE.Rmd

if (!("limma" %in% installed.packages())) {
  # Install limma
  BiocManager::install("limma", update = FALSE)
}

if (!("edgeR" %in% installed.packages())) {
  # Install limma
  BiocManager::install("edgeR", update = FALSE)
}
```

Load libraries.

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Attach libraries
library(limma)
library(edgeR)
```

Create output folders.

```{r}
# Create the results folder if it doesn't exist
if (!dir.exists("results")) {
  dir.create("results")
}

# Create the plots folder if it doesn't exist
if (!dir.exists("plots")) {
  dir.create("plots")
}
```

### Read in refine.bio data

```{r Import data from .tsv files}
# Read in metadata tsv file
metadata.df <- readr::read_tsv(file.path("data", "metadata_SRP078441.tsv")) %>%
  # Get rid of columns that have only NAs
  dplyr::select(-which(apply(is.na(.), 2, all)))

# Read in data tsv file
expression.df <- readr::read_tsv(file.path("data", "SRP078441.tsv")) %>%
  tibble::column_to_rownames("Gene") # Store gene names as the rownames for now
```

The information we need to make the comparison is in the `refinebio_title` 
column of the metadata data.frame.

```{r}
head(metadata.df$refinebio_title)
```

> We'll use some tidyverse functionality to extract this information into it's

Reviewer: ```suggestion
We'll use some tidyverse functionality to extract this information into its
```

own column to make it easier to use.

```{r}
metadata.df <- metadata.df %>%
  # the last bit of the title, separated by "-" contains the mutation 
  # information that we want to extract
  dplyr::mutate(mutation_status = stringr::word(refinebio_title,
                                                -1, sep = "-")) %>%
  # now let's make this simpler
  dplyr::mutate(mutation_status = dplyr::case_when(
    grepl("ASXL1|ASXL2", mutation_status) ~ "mutation",
    grepl("ASXLwt", mutation_status) ~ "wildtype"
  )) 
```

Let's take a look at `metadata.df` to see if this worked.

```{r}
# looking at the first 6 rows of the metadata_df and only at the columns that
# contain the title and the mutation status we extracted from the title
head(dplyr::select(metadata.df, refinebio_title, mutation_status))
```

## Differential expression analysis

Now that we've prepped our data, we're ready to test for differentially 
expressed genes.

> Let's make sure samples are in the same order.

Reviewer: ```suggestion
Let's check if the samples are in the same order.
```

```{r}
all.equal(colnames(expression.df), metadata.df$refinebio_accession_code)
```

```{r}

> # put samples in ascending orderaccording to their accession code in the 

Reviewer: ```suggestion
# put samples in ascending order according to their accession code in the 
```

# metadata
metadata.df <- metadata.df %>%
  dplyr::arrange(refinebio_accession_code)

# use the metadata.df information to reorder the columns of the expression
# data frame
expression.df <- expression.df[, metadata.df$refinebio_accession_code]
```

Check again.

```{r}
all.equal(colnames(expression.df), metadata.df$refinebio_accession_code)
```

### limma-voom

Here, we're following the steps in [this section](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom) 
of the [tximport vignette](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html).

> ```{r}

Reviewer: This chunk feels like it has a lot of steps at once as far as conceptually keeping track of what is happening to the data at each. Can we split this into some smaller chunks for easier bites?

Reviewer: What I'm saying is if we just turn some of your comments into notebook text it gives people more of a chance to digest each chunk before going to the next one and also makes them slow down which helps understanding I think.

# construct a DGEList object from the refine.bio output

> y <- DGEList(as.matrix(expression.df))

Reviewer: ```suggestion
y <- DGEList(as.matrix(expression.df))
```
I get that you are just copying from the tximport vignette, but why did they name it `y`? That doesn't tell me anything about the object. Can we call it something that explains more about what it is?

Reviewer: Or if we can explain why we are naming this object `y`.

> # filtering

Reviewer: ```suggestion
# filtering so we only keep genes that have sufficiently large counts
```

keep <- filterByExpr(y)
y <- y[keep, ]

# edgeR calculation of normalization factors to scale raw library sizes
y <- calcNormFactors(y)

# design matrix that captures the mutated vs. wild type comparison
design <- model.matrix(~ metadata.df$mutation_status)

# this is the step that transforms data to log2CPM, estimates the mean-variance
# relationship and computes observation-level weights

> v <- voom(y, design)  # ready for lmFit

Reviewer: ```suggestion
v <- voom(y, design)  # ready for lmFit
```
Can we change this object name too?

```

```{r Apply linear model}
# Apply linear model to data
fit <- lmFit(v, design = design)

# Apply empirical Bayes to smooth standard errors
fit <- eBayes(fit)

# Apply multiple testing correction and obtain stats
stats <- topTable(fit, number = nrow(expression.df)) %>% 
  tibble::rownames_to_column("Gene")
```

Write the `topTable` output to file.

```{r  Write results}
readr::write_tsv(stats, file.path("results", "SRP078441_limma_results.tsv"))
```

### Volcano plot

```{r}
# Create volcano plot 
voc.plot <- volcanoplot(fit, main = "Volcano Plot SRP078441")

# Save plot to png
png(file.path("plots", "volcano_plot_SRP078441.png"))
volcanoplot(fit, main = "Volcano Plot SRP078441")
dev.off()

> # Print out the plot here

Reviewer: This printing out part that doesn't actually print out is my bad but I think it will also be unnecessary here.

expressed genes.

Let's check if samples are in the same order.

```{r}
all.equal(colnames(expression.df), metadata.df$refinebio_accession_code)
```

```{r}
# put samples in ascending order according to their accession code in the 
# metadata
metadata.df <- metadata.df %>%
  dplyr::arrange(refinebio_accession_code)

# use the metadata.df information to reorder the columns of the expression
# data frame
expression.df <- expression.df[, metadata.df$refinebio_accession_code]
```

Check again.

```{r}
all.equal(colnames(expression.df), metadata.df$refinebio_accession_code)
```

### limma-voom

Here, we're following the steps in [this section](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom) 
of the [tximport vignette](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html).

To use `edgeR` functionality such as filtering and calculating normalization
factors, we will construct a [`DGEList`](https://www.rdocumentation.org/packages/edgeR/versions/3.10.5/topics/DGEList-class) 
using the refine.bio data as the `counts`.

```{r}
dge.list <- DGEList(as.matrix(expression.df))
```

> `DGEList` is a particular type of S4 class.

Reviewer: ```suggestion
`DGEList` is a particular type of [S4 class](http://adv-r.had.co.nz/S4.html).
```
I don't know that `S4 class` will be particularly meaningful to this audience, but I could be misunderstanding what audience we are expecting.

Reviewer: On second thought, keep the S4 comment, but let's just link it to Hadley's chapter.

```{r}
str(dge.list)
```

Filter to only retain genes with enough counts for statistical analysis.

```{r}
keep <- filterByExpr(dge.list)
dge.list <- dge.list[keep, ]
```

If we check the structure again, we can see many fewer genes are retained.

```{r}
str(dge.list)
```

The next step is to calculate normalization factors for scaling the raw 
library sizes.

```{r}
dge.list <- calcNormFactors(dge.list)
```

If we take another look at the structure of `dge.list`, we can see that the

> `norm.factors` element has changed as a result of running `calcNormFactors`.å

Reviewer: ```suggestion
`norm.factors` element has changed as a result of running `calcNormFactors`.
```
